tests/test_hubmixin.py

```diff
@@ -91,6 +94,57 @@ def test_save_pretrained(self):
         self.assertTrue("pytorch_model.bin" in files)
         self.assertEqual(len(files), 2)
 
+    def test_save_pretrained_with_push_to_hub(self):
+        REPO_NAME = repo_name("save")
+        save_directory = f"{WORKING_REPO_DIR}/{REPO_NAME}"
+        config = {"hello": "world"}
+        model = DummyModel()
+        model.push_to_hub = Mock()
+        model._save_pretrained = Mock()  # disable _save_pretrained to speed-up
+
+        # Not pushed to hub
+        model.save_pretrained(save_directory)
+        model.push_to_hub.assert_not_called()
+
+        # Push to hub with repo_id
+        model.save_pretrained(
+            save_directory, push_to_hub=True, repo_id="CustomID", config=config
+        )
+        model.push_to_hub.assert_called_with(repo_id="CustomID", config=config)
+
+        # Push to hub with default repo_id (based on dir name)
+        model.save_pretrained(save_directory, push_to_hub=True, config=config)
+        model.push_to_hub.assert_called_with(repo_id=REPO_NAME, config=config)
+
+        # Push to hub with deprecated kwargs (git-based)
+        model.save_pretrained(
+            save_directory,
+            push_to_hub=True,
+            config=config,
+            repo_path_or_name="custom_repo_name",
+            git_email="myemail",
+            git_user="gituser",
+        )
+        model.push_to_hub.assert_called_with(
+            repo_path_or_name="custom_repo_name",
+            config=config,
+            git_email="myemail",
+            git_user="gituser",
+        )
+
+        # Push to hub with deprecated kwargs + use default repo_name + no config
+        model.save_pretrained(
+            save_directory,
+            push_to_hub=True,
+            git_email="myemail",
+            git_user="gituser",
+        )
+        model.push_to_hub.assert_called_with(
+            repo_path_or_name=save_directory,
+            git_email="myemail",
+            git_user="gituser",
+        )
```

Comment:
Shouldn't these raise warnings? Why not use `expect_deprecation` for these? (also a bonus point for splitting the tests)

Reply:
@osanseviero Hmm, yes and no. Here I'm testing how `save_pretrained` propagate keyword arguments to `push_to_hub` in 5 different cases. Since I did not want to test `push_to_hub` itself I just used a mock for it to check which arguments are passed. Since `push_to_hub` itself is never run in the test, the deprecation warnings are not shown. To make it more explicit, I renamed `model` to `mocked_model`.

About splitting tests, I get the point what checking different cases but I didn't know how to do it. A basic way would be to duplicate init code for each test but I'm not a huge fan:

```py
def test_case_xxx(self):
    # next 6 lines for init
    REPO_NAME = repo_name("save")
    save_directory = f"{WORKING_REPO_DIR}/{REPO_NAME}"
    config = {"hello": "world"}
    mocked_model = DummyModel()
    mocked_model.push_to_hub = Mock()
    mocked_model._save_pretrained = Mock()

    mocked_model.save_pretrained(...) # <- actual code that changes for each case
    mocked_model.push_to_hub.assert_called_with(...) # <- actual code that changes for each case
```
I could also group the first 6 lines in a separate function but it would make the test less readable.

In the end, that's why I kept all the cases together (also the fact that they don't have any external side effect since it's all mocked) but I'd be happy to change if you have an opinion/suggestion about it.

Reply:
TL;DR: I am open to suggestion on how to make this clean 😇